---
title: 'CA2200: повторный вызов для сохранения сведений о стеке (анализ кода)'
description: 'Дополнительные сведения о правиле анализа кода CA2200: Rethrow для сохранения сведений о стеке'
ms.date: 11/04/2016
ms.topic: reference
f1_keywords:
- RethrowToPreserveStackDetails
- CA2200
helpviewer_keywords:
- CA2200
- RethrowToPreserveStackDetails
author: gewarren
ms.author: gewarren
dev_langs:
- CSharp
- VB
ms.openlocfilehash: c49f3dbab810ce32bfab2d13fbbe9c9f79da6df5
ms.sourcegitcommit: ddf7edb67715a5b9a45e3dd44536dabc153c1de0
ms.translationtype: MT
ms.contentlocale: ru-RU
ms.lasthandoff: 02/06/2021
ms.locfileid: "99719933"
---
# <a name="ca2200-rethrow-to-preserve-stack-details"></a><span data-ttu-id="c2e63-103">CA2200. Повторно порождайте исключения для сохранения сведений стека</span><span class="sxs-lookup"><span data-stu-id="c2e63-103">CA2200: Rethrow to preserve stack details</span></span>

| | <span data-ttu-id="c2e63-104">Значение</span><span class="sxs-lookup"><span data-stu-id="c2e63-104">Value</span></span> |
|-|-|
| <span data-ttu-id="c2e63-105">**Идентификатор правила**</span><span class="sxs-lookup"><span data-stu-id="c2e63-105">**Rule ID**</span></span> |<span data-ttu-id="c2e63-106">CA2200</span><span class="sxs-lookup"><span data-stu-id="c2e63-106">CA2200</span></span>|
| <span data-ttu-id="c2e63-107">**Категория**</span><span class="sxs-lookup"><span data-stu-id="c2e63-107">**Category**</span></span> |[<span data-ttu-id="c2e63-108">Использование</span><span class="sxs-lookup"><span data-stu-id="c2e63-108">Usage</span></span>](usage-warnings.md)|
| <span data-ttu-id="c2e63-109">**Исправление не критическое или не критическое**</span><span class="sxs-lookup"><span data-stu-id="c2e63-109">**Fix is breaking or non-breaking**</span></span> |<span data-ttu-id="c2e63-110">Не критическое</span><span class="sxs-lookup"><span data-stu-id="c2e63-110">Non-breaking</span></span>|

## <a name="cause"></a><span data-ttu-id="c2e63-111">Причина</span><span class="sxs-lookup"><span data-stu-id="c2e63-111">Cause</span></span>

<span data-ttu-id="c2e63-112">Возникает повторное исключение, и в инструкции явно указывается исключение `throw` .</span><span class="sxs-lookup"><span data-stu-id="c2e63-112">An exception is rethrown and the exception is explicitly specified in the `throw` statement.</span></span>

## <a name="rule-description"></a><span data-ttu-id="c2e63-113">Описание правила</span><span class="sxs-lookup"><span data-stu-id="c2e63-113">Rule description</span></span>

<span data-ttu-id="c2e63-114">После возникновения исключения часть информации, которую он передает, является трассировкой стека.</span><span class="sxs-lookup"><span data-stu-id="c2e63-114">Once an exception is thrown, part of the information it carries is the stack trace.</span></span> <span data-ttu-id="c2e63-115">Трассировка стека — это список иерархии вызовов методов, который начинается с метода, вызывающего исключение, и завершается методом, перехватывающим исключение.</span><span class="sxs-lookup"><span data-stu-id="c2e63-115">The stack trace is a list of the method call hierarchy that starts with the method that throws the exception and ends with the method that catches the exception.</span></span> <span data-ttu-id="c2e63-116">Если исключение создается повторно путем указания исключения в `throw` инструкции, трассировка стека перезапускается в текущем методе, а список вызовов методов между исходным методом, вызвавшим исключение, и текущим методом теряется.</span><span class="sxs-lookup"><span data-stu-id="c2e63-116">If an exception is re-thrown by specifying the exception in the `throw` statement, the stack trace is restarted at the current method and the list of method calls between the original method that threw the exception and the current method is lost.</span></span> <span data-ttu-id="c2e63-117">Для сохранения исходных данных трассировки стека с исключением следует использовать оператор `throw` без указания исключения.</span><span class="sxs-lookup"><span data-stu-id="c2e63-117">To keep the original stack trace information with the exception, use the `throw` statement without specifying the exception.</span></span>

## <a name="how-to-fix-violations"></a><span data-ttu-id="c2e63-118">Устранение нарушений</span><span class="sxs-lookup"><span data-stu-id="c2e63-118">How to fix violations</span></span>

<span data-ttu-id="c2e63-119">Чтобы устранить нарушение этого правила, воссоздайте исключение без явного указания исключения.</span><span class="sxs-lookup"><span data-stu-id="c2e63-119">To fix a violation of this rule, rethrow the exception without specifying the exception explicitly.</span></span>

## <a name="when-to-suppress-warnings"></a><span data-ttu-id="c2e63-120">Когда следует подавлять предупреждения</span><span class="sxs-lookup"><span data-stu-id="c2e63-120">When to suppress warnings</span></span>

<span data-ttu-id="c2e63-121">Для этого правила отключать вывод предупреждений не следует.</span><span class="sxs-lookup"><span data-stu-id="c2e63-121">Do not suppress a warning from this rule.</span></span>

## <a name="example"></a><span data-ttu-id="c2e63-122">Пример</span><span class="sxs-lookup"><span data-stu-id="c2e63-122">Example</span></span>

<span data-ttu-id="c2e63-123">В следующем примере показан метод, `CatchAndRethrowExplicitly` который нарушает правило, и метод, `CatchAndRethrowImplicitly` который удовлетворяет правилу.</span><span class="sxs-lookup"><span data-stu-id="c2e63-123">The following example shows a method, `CatchAndRethrowExplicitly`, which violates the rule and a method, `CatchAndRethrowImplicitly`, which satisfies the rule.</span></span>

:::code language="csharp" source="snippets/csharp/all-rules/ca2200.cs" id="snippet1":::

:::code language="vb" source="snippets/vb/all-rules/ca2200-rethrow-to-preserve-stack-details_1.vb":::
